### PR TITLE
New acceptance criteria

### PR DESCRIPTION
**Pull request by leonlan**
# Issue #75 

## TODOs
- [x] RandomWalk
- [x] LateAcceptanceHillClimbing 
- [x] ThresholdAcceptance
- [x] GreatDeluge
- [x] NonLinearGreatDeluge
- [x] WorseAccept

### CODE REVIEW

**Comment by leonlan** [comment]
# Questions
- LateAcceptanceHillClimbing
  - New class because Santini did it also like this. However, it should be easy to adapt HillClimbing so that it can be a single class. What do you think? (Also relevant question for GreatDeluge and NonLinearGreatDeluge).
  - Semantics: "Better" in this case means "not worsening" the solution, which is the terminology used by Santini. Change or keep?
- General questions
  - I tried making some of the tests more readible by using `None` for rnd/curr/best/cand if not used. Helps in particular when reading tests for RTRT and TA I think. Not sure if bad practice.

**Comment by N-Wouda** [comment]
Thanks for the PR @leonlan! 

I'm off to a conference today and tomorrow. I'll try to have a look late tomorrow!

File: alns/accept/ThresholdAcceptance.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    ----------
+    [1]: Santini, A., Ropke, S. & Hvattum, L.M. A comparison of acceptance
+         criteria for the adaptive large neighbourhood search metaheuristic.
+         *Journal of Heuristics* (2018) 24 (5): 783–815.
+    [2]: Dueck, G., Scheuer, T. Threshold accepting: A general purpose
+         optimization algorithm appearing superior to simulated annealing.
+         *Journal of Computational Physics* (1990) 90 (1): 161-175.
+    """
+
+    def __call__(self, rnd, best, current, candidate):
```
Comment (N-Wouda) [inline alns/accept/ThresholdAcceptance.py:38]: Question:

This looks a lot like the implementation of RRT to me. What's different?

There's this 1993 paper about RRT, again by Dueck, where they also introduce great deluge. That's perhaps where Santini got his normalisation from, because the paper talks about some "record-deviation" parameter.

**Comment by N-Wouda** [comment]
> New class because Santini did it also like this. However, it should be easy to adapt HillClimbing so that it can be a single class. What do you think? (Also relevant question for GreatDeluge and NonLinearGreatDeluge).

I think we should prefer to extend existing classes. Santini describes them all separately, but code wise they're so related it makes sense to expand existing classes. I will comment on this more as we implement more (variations on) criteria.

File: alns/accept/ThresholdAcceptance.py (hunk 1 above)
Comment (leonlan) [reply alns/accept/ThresholdAcceptance.py:38]: RRT compares `candidate` against `best`, whereas TA compares it against `current`.

- I noticated that the paper by Dueck and Scheuer (1990) actually introduced TA. The paper in by Dueck in (1993) introduces RRT. 
- The "record-deviation" parameter in Dueck (1993) is a threshold value - similar to what the current implementation does. 
- Santini describes in the paper also that a "threshold starts at a large value and decreases at every iteration", which, to me, seems like his intention was to implement it using an absolute gap threshold. 
- If I had to take a guess based on my own experiences, I usually infer the absolute gap thresholds from the relative deviation, i.e., the initial threshold is 5% of the initial objective value. So directly implementing a threshold based on relative gap makes actually quite a lot of sense from a user POV.

**Comment by leonlan** [comment]
I have finally implemented all criteria presented in Santini et al. (2018). 😄 

## List of criteria
- [x] RandomWalk: in `RandomWalk`
- [x] LateAcceptanceHillClimbing: in `HillClimbing`
- [x] ThresholdAcceptance: in `ThresholdAcceptance`
- [x] GreatDeluge: in `GreatDeluge`
- [x] NonLinearGreatDeluge: in `GreatDeluge`
- [x] WorseAccept: in `WorseAccept`

## Open questions
- **Merging TA and RRT**: Similar to GD and HC, we could merge TA and RRT into a single criterion class because they are so similar. But because the names are quite different, it might actually create confusion. 
- **Semantics**: In RRT and TA, "better" means not worsening the solution as per Santini et al. (2018). Should we be more consistent and use the terminology that a solution is better iff it has strictly lower objective value? And use "not worsen" or "not deteriorate" iff it has equal or lower objective value?
-  **Readability of tests**: I tried making some of the tests more readible by using `None` for `rnd/curr/best/cand` if not used.  I haven't been very consistent everywhere, but paid specific attention if I think helped reading tests (for RRT and TA specifically).

File: alns/accept/tests/test_hill_climbing.py [hunk 2]
```diff
@@ -1,30 +1,92 @@
...
+    """
+    lahc = HillClimbing(on_current=False, n_last=n_last)
+
+    def accept(current, candidate):
+        return lahc(None, None, current, candidate)
+
+    accept(Two(), One())
+
+    for _ in range(n_last - 1):
+        accept(Zero(), One())
```
Comment (N-Wouda) [inline alns/accept/tests/test_hill_climbing.py:63]: Refactor: these are testable statements, but are not currently tested.

Comment (N-Wouda) [inline alns/accept/tests/test_hill_climbing.py:63]: Refactor: these are testable statements, but are not currently tested.

File: alns/accept/tests/test_hill_climbing.py [hunk 3]
```diff
@@ -1,30 +1,92 @@
...
     Tests if the hill climbing method accepts a solution that results in the
     same objective value.
     """
     hill_climbing = HillClimbing()
     assert_(hill_climbing(rnd.RandomState(), Zero(), Zero(), Zero()))
+
+
+"""
+Late Acceptance Hill Climbing
+"""
```
Comment (N-Wouda) [inline alns/accept/tests/test_hill_climbing.py:40]: Comment: I would add this to the test names (e.g. `def test_lhac_...()`). You already do that for regular hill climbing just before, so it seems more consistent.

Comment (N-Wouda) [inline alns/accept/tests/test_hill_climbing.py:40]: Comment: I would add this to the test names (e.g. `def test_lhac_...()`). You already do that for regular hill climbing just before, so it seems more consistent.

File: alns/accept/tests/test_great_deluge.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    # hence Two should be rejected
+    assert_(not great_deluge(None, One(), Zero(), Two()))
+
+
+def test_non_linear_accepts_improving_current():
+    great_deluge = GreatDeluge(2, 0.5, 1, 1, method="non-linear")
+
+    # Two does not improve over the threshold (2) but does improve over the
+    # current solution value (3), hence One should be accepted
+    assert_(great_deluge(None, One(), Three(), Two()))
```
Comment (N-Wouda) [inline alns/accept/tests/test_great_deluge.py:141]: Question: can we also test this without another mock state?

Comment (N-Wouda) [inline alns/accept/tests/test_great_deluge.py:141]: Question: can we also test this without another mock state?

File: alns/accept/tests/test_great_deluge.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    great_deluge = GreatDeluge(1.5, 0.1, 0.02, 1, method="non-linear")
+
+    # The first candidate is rejected and linearly increases the threshold
+    great_deluge(None, One(), Zero(), Two())
+    assert_equal(great_deluge._threshold, 1.51)
+
+    # The second candidate is accepted and exponentially decreases the threshold
+    great_deluge(None, One(), Zero(), Zero())
+    new_threshold = 1.51 * np.exp(-1) + 1
+    assert_equal(great_deluge._threshold, new_threshold)
```
Comment (N-Wouda) [inline alns/accept/tests/test_great_deluge.py:118]: Refactor: this is testing the implementation, by directly accessing the `_threshold` parameter. Don't do that.

Instead, test outcomes: you know how things should be updated, so you can construct solutions that are accepted/rejected given what the threshold should be.

Comment (N-Wouda) [inline alns/accept/tests/test_great_deluge.py:118]: Refactor: this is testing the implementation, by directly accessing the `_threshold` parameter. Don't do that.

Instead, test outcomes: you know how things should be updated, so you can construct solutions that are accepted/rejected given what the threshold should be.

Comment (N-Wouda) [reply alns/accept/tests/test_great_deluge.py:118]: I think the test cases for SA do something similar.

Comment (N-Wouda) [reply alns/accept/tests/test_great_deluge.py:118]: I think the test cases for SA do something similar.

**Review by N-Wouda — CHANGES_REQUESTED** [review]

**Comment by N-Wouda** [comment]
I will have a better look soon, these are just some first thoughts as I skimmed through the tests.

File: alns/accept/tests/test_great_deluge.py (hunk 4 above)
Comment (N-Wouda) [reply alns/accept/tests/test_great_deluge.py:141]: Or maybe we should make one mock State with a variable objective:
```python
from numbers import Number


@dataclass
class VarObj(State):
    obj: Number

    def objective(self) -> Number:
        return self.obj
```
And then
```python
Zero = VarObj(0)
One = VarObj(1)
Two = VarObj(2)
```
for backwards compatiblity, but otherwise make whatever mock object that's needed available.

Comment (leonlan) [reply alns/accept/tests/test_great_deluge.py:141]: > Question: can we also test this without the new mock state `Three`?

I don't think so: in non-linear great deluge, I can't use `Zero` because of division by zero. So that leaves me with two states, which are not enough states to test with.

> Or maybe we should make one mock State with a variable objective:
> 
> ```python
> from numbers import Number
> 
> 
> class VarObj(State):
>     """Test solution state object with variable objective."""
>     
>     def __init__(self, obj: Number):
>         self.obj = obj
> 
>     def objective(self) -> Number:
>         return self.obj
> ```
> 
> And then
> 
> ```python
> Zero = VarObj(0)
> One = VarObj(1)
> Two = VarObj(2)
> ```
> 
> for backwards compatiblity, but otherwise make whatever mock object that's needed available.

For backwards compatibility we need Zero, One, Two, to be callable right? How about we leave Zero, One, Two as is, and replace `Three` with `VarObj`?

Comment (N-Wouda) [reply alns/accept/tests/test_great_deluge.py]: >For backwards compatibility we need Zero, One, Two, to be callable right? How about we leave Zero, One, Two as is, and replace Three with VarObj?

IIRC, they are just `alns.State` objects. Those have an `objective()` member method, but not much else. They are not required to be callable - other than via their `objective()` member.

File: alns/accept/HillClimbing.py [hunk 6]
```diff
@@ -1,11 +1,58 @@
...
-    that result in a worse objective value.
+    The Hill Climbing (HC) criterion only accepts progressively better solutions,
+    discarding those that result in a worse objective value.
+
+    There are three variants of the HC criterion:
+    1) Classic (HC): only accepts a candidate solution if it is better
+       than the current solution.
+    2) Late Acceptance (LAHC): only accepts a candidate solution if it is
+       better than the current solution from a number of iterations ago.
+    3) Improved LA (ILAHC): this variant combines HC and LAHC.
```
Comment (N-Wouda) [inline alns/accept/HillClimbing.py:17]: Question: how does it combine the two? [we should probably explain that]

File: alns/accept/WorseAccept.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    @property
+    def method(self) -> str:
+        return self._method
+
+    def __call__(self, rnd, best, current, candidate):
+        result = (
+            candidate.objective() < current.objective()
+            or rnd.random() < self._prob
+        )
```
Comment (N-Wouda) [inline alns/accept/WorseAccept.py:79]: I find this difficult to parse due to the split expression. Maybe refactor?
```python
res = cand.obj() < curr.obj() # always accept if better

if not res:  # maybe accept
  res = rnd.random() < self._prob
```

File: alns/accept/GreatDeluge.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                result = self._threshold + self._gamma * abs(
+                    candidate.objective() - self._threshold
+                )
+            else:
+                result = (
+                    self._threshold * math.exp(-self._delta * best.objective())
+                    + best.objective()
+                )
+
+        return result
```
Comment (N-Wouda) [inline alns/accept/GreatDeluge.py:141]: My first impression of this is "wow, that's complex!". I suspect we can simplify this quite a bit.

File: alns/accept/GreatDeluge.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        if alpha <= 1:
+            raise ValueError("Alpha must be larger than 1.")
+
+        if beta <= 0 or beta >= 1:
+            raise ValueError("Beta must be in (0, 1).")
+
+        if (
+            gamma is None or delta is None or gamma <= 0 or delta <= 0
+        ) and method == "non-linear":
+            raise ValueError("Gamma and delta must be non-negative.")
```
Comment (N-Wouda) [inline alns/accept/GreatDeluge.py:69]: Question: these parameters are optional with the linear method, right?

Refactor: I think we should split this class up. Every member method branches on whether it is the linear/non-linear method. There's very benefit we're getting from having one GreatDeluge class.

I propose we do a `LinearGreatDeluge` and `NonLinearGreatDeluge` class. Those should then be rather straightforward.

**Review by N-Wouda — CHANGES_REQUESTED** [review]

File: alns/accept/ThresholdAcceptance.py (hunk 1 above)
Comment (leonlan) [reply alns/accept/ThresholdAcceptance.py:39]: Good idea!

File: alns/accept/tests/test_great_deluge.py (hunk 4 above)
Comment (leonlan) [reply alns/accept/tests/test_great_deluge.py]: Sorry I misunderstood the proposed suggestion. 

If we rename all references of `Zero()` to `VarObj(0)` etc., do you feel that the code is easier to read? I'm a bit split about this one. Seeing `0` instead of `Zero` makes it easier to see immediately what is evaluated/compared, but there's also something "ugly" about seeing `VarObj` repeated multiple times.
```python
record_travel(rnd.RandomState(), VarObj(0), VarObj(0), VarObj(1))
```

File: alns/accept/HillClimbing.py [hunk 10]
```diff
@@ -1,11 +1,58 @@
...
+
+    Parameters
+    ----------
+    on_current: bool
+        A bool to denote whether to accept candidate solutions if they are
+        better than the current solution. Default: True.
```
[the remark this answers is not in the document]
Comment (leonlan) [reply alns/accept/HillClimbing.py:23]: I think it's a reasonable default as well, but the entire premise of "Improved" LAHC is that it improves over LAHC by having this addition. The reference mentioned in Santini et al. (2018) is by [Burke and Bykov (2012)](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.261.3472&rep=rep1&type=pdf).

File: alns/accept/tests/test_great_deluge.py (hunk 4 above)
Comment (N-Wouda) [reply alns/accept/tests/test_great_deluge.py]: Oh! I see what you mean now.

In `states.py` we could do the following instead as drop-in replacements for now:
```python
Zero = lambda: VarObj(0)
One = lambda: VarObj(1)
Two: lambda: VarObj(2)
```
At some later point we can then simplify by taking out the lambdas, but I do not want to burden this PR with that change.

File: alns/accept/HillClimbing.py [hunk 11]
```diff
@@ -1,11 +1,58 @@
...
+    n_last: int
+        A non-negative integer that specifies the number of last current
+        solutions that need to be stored. Default: None, which means that no
+        previous current solutions are stored.
```
[the remark this answers is not in the document]
Comment (leonlan) [reply alns/accept/HillClimbing.py:27]: Yes it can be zero, `deque` can be initialized with `maxlen=0` as well.

File: alns/accept/HillClimbing.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (leonlan) [reply alns/accept/HillClimbing.py:23]: My suggestion is to make another class `LateAcceptanceHillClimbing`, whose implementation follows the one proposed by Burke and Bykov (2017). 
1. LAHC: default, only accept candidate if improving L-th previous current
2. ILAHC: LAHC but also accepts if candidate improves current
3. (no name): LAHC with better history recording, i.e., only improving previous currents are stored. This feels to me as a history based on L-th previous best.

File: alns/accept/GreatDeluge.py (hunk 9 above)
Comment (leonlan) [reply alns/accept/GreatDeluge.py:69]: Answer: that's correct.

Refactoring is a good idea. I prefer to name it `GreatDeluge`, which is how the criterion's name as known in the literature.

Comment (N-Wouda) [reply alns/accept/GreatDeluge.py:69]: Agreed! I was only thinking from the code, but if the literature calls it GreatDeluge we should probably do so too.

What's the non-linear variant called? Does that have a name other than non-linear great deluge?

**Comment by leonlan** [comment]
@N-Wouda Should I make separate PRs for each group of criteria? The discussion is hard to follow currently 😅

File: alns/accept/GreatDeluge.py (hunk 9 above)
Comment (leonlan) [reply alns/accept/GreatDeluge.py:69]: The [original paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.723.2759&rep=rep1&type=pdf) calls it "Non-Linear Great Deluge". 

I haven't looked for enough great deluge papers to know if there are other names, but Santini et al. (2018) call it NLGD as well and searching for non-linear great deluge on [Google scholar ](
https://scholar.google.com/scholar?hl=nl&as_sdt=0%2C5&q=non+linear+great+deluge&btnG=)seems to give enough matches.

File: alns/accept/HillClimbing.py (hunk 6 above)
Comment (leonlan) [reply alns/accept/HillClimbing.py:17]: Answer:
```
1) HillClimbing()
2) HillClimbing(False, some_number)
3) HillClimbing(n_last=some_number)
```
But we should probably change this. See https://github.com/N-Wouda/ALNS/pull/76#discussion_r889680149.

**Comment by N-Wouda** [comment]
> @N-Wouda Should I make separate PRs for each group of criteria? The discussion is hard to follow currently 😅

That'd certainly make reviewing things much easier as well! This has gotten quite big :-).

**Comment by leonlan** [comment]
> > @N-Wouda Should I make separate PRs for each group of criteria? The discussion is hard to follow currently 😅
> 
> That'd certainly make reviewing things much easier as well! This has gotten quite big :-).

Alright, I made the first steps. I won't be able to work on it anymore today, but I'll continue tomorrow!

**Comment by N-Wouda** [comment]
@leonlan 

>Alright, I made the first steps. I won't be able to work on it anymore today, but I'll continue tomorrow! Some of the PRs I've made to drafts, the others are ready for review again.

Great! I'll have another look today or tomorrow, but after that I'll be off to (yet another) conference for a week, and my responses might be a bit slow from 7/6 to 14/6.

**Comment by N-Wouda** [comment]
I am closing this PR since we will continue the discussion in the other PRs.
